### PR TITLE
feat: Users other than `self` should have the permission field in the user data

### PR DESCRIPTION
**Pull request by makingthematrix**
## What's new in this PR?

The functionality for filtering out from the user search results which should not be visible for partners. The results visible for standard team members are also slightly changed, taking into account that some of the results are partners.

For details please refer to the documentation:
https://github.com/wearezeta/documentation/blob/master/topics/teams/use%20cases/clients/101-search-for-partners.md

### CODE REVIEW

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 1]
```diff
@@ -67,34 +79,48 @@ class UserSearchService(selfUserId:           UserId,
...
-    searchLocal(filter).map(_.filter(u => !(u.isGuest(teamId) && teamOnly)))
+  private lazy val isPartner = userPrefs(SelfPermissions).apply()
+    .map(decodeBitmask)
+    .map(ps => PartnerPermissions.subsetOf(ps) && PartnerPermissions.size == ps.size)
+
+  private def filterForPartner(query: Filter, searchResults: IndexedSeq[UserData]): Future[IndexedSeq[UserData]] = {
+    lazy val knownUsers = membersStorage.getByUsers(searchResults.map(_.id).toSet).map(_.map(_.userId).toSet)
+    isPartner.flatMap {
+      case true if teamId.isDefined =>
+        ZLog.verbose(s"filterForPartner1 Q: $query, RES: ${searchResults.map(_.getDisplayName)}) with partner = true and teamId")
```
Comment (marcoconti83) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: How does this interact with the new logging system? The display name should not go into public logs

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 2]
```diff
@@ -67,34 +79,48 @@ class UserSearchService(selfUserId:           UserId,
...
+        knownUsers.map(knownUsersIds => searchResults.filter(u => knownUsersIds.contains(u.id)))
+      case false if teamId.isDefined =>
+        ZLog.verbose(s"filterForPartner2 Q: $query, RES: ${searchResults.map(_.getDisplayName)}) with partner = false and teamId")
```
Comment (marcoconti83) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: Log

File: zmessaging/src/main/scala/com/waz/service/teams/TeamsService.scala [hunk 3]
```diff
@@ -64,6 +64,15 @@ class TeamsServiceImpl(selfUser:           UserId,
 
   private implicit val dispatcher = SerialDispatchQueue()
 
+  private val shouldSyncTeam = userPrefs.preference(UserPreferences.ShouldSyncTeam)
+
+  for {
+    shouldSync <- shouldSyncTeam()
+  } if (shouldSync && teamId.isDefined) {
+    verbose(l"Syncing the team $teamId")
```
Comment (marcoconti83) [inline zmessaging/src/main/scala/com/waz/service/teams/TeamsService.scala]: Log

File: zmessaging/src/main/scala/com/waz/sync/client/TeamsClient.scala [hunk 4]
```diff
@@ -113,7 +119,10 @@ object TeamsClient {
 
   case class TeamMembers(members: Seq[TeamMember])
 
-  case class TeamMember(user: UserId, permissions: Option[Permissions])
+  case class TeamMember(user: UserId, permissions: Option[Permissions], created_by: Option[UserId])
+
+  implicit val TeamMemberLogShow: LogShow[TeamMember] =
+    LogShow.create(tm => s"TeamMember(${tm.user}, ${tm.permissions}, ${tm.created_by})")
```
Comment (marcoconti83) [inline zmessaging/src/main/scala/com/waz/sync/client/TeamsClient.scala]: Log leaking user ID?

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala (hunk 1 above)
Comment (marcoconti83) [reply zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: From 1:1 discussion: it won't, `ZLog.verbose` only works if compiling in debug

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala (hunk 2 above)
Comment (marcoconti83) [reply zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: See above for resolution

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 5]
```diff
@@ -67,34 +79,48 @@ class UserSearchService(selfUserId:           UserId,
   private val exactMatchUser = new SourceSignal[Option[UserData]]()
   private val signalMap = mutable.HashMap[SearchQuery, Signal[IndexedSeq[UserData]]]()
 
-  def usersForNewConversation(filter: Filter = "", teamOnly: Boolean): Signal[IndexedSeq[UserData]] =
-    searchLocal(filter).map(_.filter(u => !(u.isGuest(teamId) && teamOnly)))
+  private lazy val isPartner = userPrefs(SelfPermissions).apply()
+    .map(decodeBitmask)
+    .map(ps => PartnerPermissions.subsetOf(ps) && PartnerPermissions.size == ps.size)
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: see comment above about this maybe being simplified to simple calling `.equals()`

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 6]
```diff
@@ -67,34 +79,48 @@ class UserSearchService(selfUserId:           UserId,
...
+        }
+      case _ => Future.successful(searchResults)
+    }
+  }
+
+  def usersForNewConversation(filter: Filter = "", teamOnly: Boolean): Signal[IndexedSeq[UserData]] = {
+    val isHandle = Handle.isHandle(filter)
+    searchLocal(filter, isHandle = isHandle)
+      .map(_.filter(u => !(u.isGuest(teamId) && teamOnly)))
+      .flatMap(res => Signal.future(filterForPartner(filter, res)))
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: My understanding is we are creating Signals from data which is unlikely to ever change. It feels like `usersForNewConversation` and `usersToAddToConversation` might be better off returning `Future` instead of `Signal`. Then we can get rid of these ugly `Signal.future` wraps, doing away with the subscriptions to them also. What do you think?

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 7]
```diff
@@ -225,27 +247,35 @@ class UserSearchService(selfUserId:           UserId,
...
-      local <- searchLocal(filter, showBlockedUsers = true)
-      convs <- conversations
-      dir   <- directorySearch
+      top       <- topUsers
+      isPartner <- Signal.future(isPartner)
+      local     <- searchLocal(filter, showBlockedUsers = true, isHandle = isHandle)
+      local     <- Signal.future(filterForPartner(filter, local))
+      convs     <- conversations
+      dir       <- if (isPartner) Signal.const(IndexedSeq.empty[UserData]) else directorySearch
+      dir       <- Signal.future(filterForPartner(filter, dir))
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: I think it might be nicer to merge these last two lines. I'm surprised you can reuse variable names in a for comprehension. I think a small method encapsulating these two lines and flatmapping them together would be more readable, otherwise maybe just using different variable names for both these lines? As it stands it seems a bit weird. Also again, maybe this whole method should return a `Future` if we are simply wrapping `filterForPartner` again?

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 8]
```diff
@@ -225,27 +247,35 @@ class UserSearchService(selfUserId:           UserId,
...
 
     for {
-      updated <- userService.updateUsers(results)
-      _       <- userService.syncIfNeeded(updated.map(_.id), Duration.Zero)
-      ids     = results.map(_.id)(breakOut): Vector[UserId]
-      _       = verbose(l"updateSearchResults($query, ${results.map(_.handle)})")
-      _       <- queryCache.updateOrCreate(query, updating(ids), SearchQueryCache(query, clock.instant(), Some(ids)))
+      userData <- usersStorage.listAll(ids)
+      filtered <- filterForPartner(query.filter, userData)
+      filteredIds = filtered.map(_.id).toSet
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: These two lines can be combined, as we never use `filtered` elsewhere

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 9]
```diff
@@ -255,21 +285,20 @@ class UserSearchService(selfUserId:           UserId,
...
         queryCache.updateOrCreate(query, updating(userId), SearchQueryCache(query, clock.instant(), Some(Vector(userId))))
-    }(Threading.Background)
+    }
 
     Future.successful({})
   }
 
   def searchUserData(query: SearchQuery): Signal[IndexedSeq[UserData]] = signalMap.getOrElseUpdate(query, returning( startNewSearch(query) ) { _ =>
-    CancellableFuture.delay(cacheExpiryTime).map { _ =>
+    CancellableFuture.delay(cacheRefreshInterval).map { _ =>
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: What's the reason behind this change? Looks like we are "expiring" instead of "refreshing" here.

File: zmessaging/src/main/scala/com/waz/sync/handler/UsersSyncHandler.scala [hunk 10]
```diff
@@ -40,7 +40,8 @@ class UsersSyncHandler(assetSync: AssetSyncHandler,
                        assets: AssetService,
                        usersClient: UsersClient,
                        imageGenerator: ImageAssetGenerator,
-                       otrSync: OtrSyncHandler) {
+                       otrSync: OtrSyncHandler
+                      ) {
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/sync/handler/UsersSyncHandler.scala]: This change doesn't seem necessary, can we roll it back?

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala (hunk 6 above)
Comment (makingthematrix) [reply zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: Yes, but that will propagate changes deep into UI which right now uses signals. After working on this I agree futures would be a better solution. After all, if I search for something, I don't really expect the results of my search to change while I'm looking at it. If I want to refresh it, I can do by simply searching again.
Unfortunately, we don't have time to rewrite it all right now. I started a separate PR for that, but it will take more time than we have.

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala (hunk 7 above)
Comment (CalumMcCall) [reply zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: We could simply rename the second instance of `dir` to `dirResults` or something. My real issue is just this reuse of variable names.

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala (hunk 9 above)
Comment (makingthematrix) [reply zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: Removed, together with the `signalMap`.

File: zmessaging/src/main/scala/com/waz/model/Handle.scala [hunk 11]
```diff
@@ -25,11 +25,11 @@ case class Handle(string: String) extends AnyVal {
   override def toString : String = string
 
   def startsWithQuery(query: String): Boolean = {
-     string.startsWith(Handle.stripSymbol(query).toLowerCase)
+    query.nonEmpty && string.startsWith(Handle.stripSymbol(query).toLowerCase)
   }
 
   def exactMatchQuery(query: String): Boolean = {
-    string == Handle.stripSymbol(query).toLowerCase
+    query.nonEmpty && string == Handle.stripSymbol(query).toLowerCase
```
[the remark this answers is not in the document]
Comment (makingthematrix) [reply zmessaging/src/main/scala/com/waz/model/Handle.scala]: `.startsWith("")` returns `true` and that's why `.nonEmpty` is needed in `startsWithQuery`, but it's not needed in `exactMatchQuery`. Fixed.

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 12]
```diff
@@ -297,14 +296,8 @@ class UserSearchService(selfUserId:           UserId,
...
-  def searchUserData(query: SearchQuery): Signal[IndexedSeq[UserData]] = signalMap.getOrElseUpdate(query, returning( startNewSearch(query) ) { _ =>
-    CancellableFuture.delay(cacheRefreshInterval).map { _ =>
-      signalMap.remove(query)
-      queryCache.remove(query)
-    }
-  })
-
-  private def startNewSearch(query: SearchQuery): Signal[IndexedSeq[UserData]] = returning( queryCache.optSignal(query) ){ _ =>
+  // not private for tests
+  def searchUserData(query: SearchQuery): Signal[IndexedSeq[UserData]] = returning( queryCache.optSignal(query) ){ _ =>
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: Was this caching code you removed not needed anymore?

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala [hunk 13]
```diff
@@ -17,7 +17,7 @@
  */
 package com.waz.service
 
-import com.waz.ZLog
+import com.waz.ZLog.verbose
```
Comment (CalumMcCall) [inline zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: Should we maybe just migrate to ZLog2? You can use `showString` to display normal strings, to save you having to create instances of LogShow

File: zmessaging/src/main/scala/com/waz/service/UserSearchService.scala (hunk 12 above)
Comment (makingthematrix) [reply zmessaging/src/main/scala/com/waz/service/UserSearchService.scala]: Some time ago we suspected that because the user search sometimes returns a lot of results, it may trigger GC  and freeze the app. One way to avoid it was to call `System.gc()` before even querying. The other was to cache the signals instead of creating new ones - that's why the `signalMap` with the expiry time for signals.
But as I test it now it seems that it does not help. I would like to remove `signalMap` (and this call to `gc`) and see what happens in Internal and Beta. If we don't get any feedback about slower search (I guess we don't) it means it's not needed anymore.
In the meantime we created a lot of batch versions for methods in services and handlers. Instead of calling `add*` or `update*` many times for each instance of a case class, we call a batch method for the whole sequence of them. It might be the reason why the cache is not really ding anything anymore.

It also shows that we lack performance tests.
